### Exploring used cars data from the German ebay website

Here we will be exploring data about used cars from the German ebay website.

In [36]:
import pandas as pd
import numpy as np

In [37]:
autos = pd.read_csv("autos.csv", encoding="latin1")

In [38]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Dataset has 50,000 rows. Columns 'vehicle type', 'gearbox', 'model', 'fueltype', 'notrepaireddamange' have some null values. All columns are either strings or 64 bit integers.

#### Rename columns using python naming convention.

In [39]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [40]:
new_col_names = {'dateCrawled': 'date_crawled', 'offerType': 'offer_type', 'abtest': 'ab_test',
       'vehicleType': 'vehicle_type', 'yearOfRegistration': 'registration_year', 'powerPS': 'power_PS', 'monthOfRegistration': 'registration_month', 'fuelType': 'fuel_type',
       'notRepairedDamage': 'unrepaired_damange', 'dateCreated': 'ad_created', 'nrOfPictures': 'nr_of_pictures', 'postalCode': 'postal_code',
       'lastSeen': 'last_seen'}

In [41]:
autos = autos.rename(columns=new_col_names)

In [42]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  object
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_PS            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damange  40171 non-null  object
 16  ad_created          50

#### Get an overview of the data.

In [43]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count              50000        50000   50000      50000  50000   50000   
unique             12073        38754       2          2   2357       2   
top     28/03/2016 14:49  Ford_Fiesta  privat    Angebot     $0    test   
freq                  15           78   49999      49999   1421   25756   
mean                 NaN          NaN     NaN        NaN    NaN     NaN   
std                  NaN          NaN     NaN        NaN    NaN     NaN   
min                  NaN          NaN     NaN        NaN    NaN     NaN   
25%                  NaN          NaN     NaN        NaN    NaN     NaN   
50%                  NaN          NaN     NaN        NaN    NaN     NaN   
75%                  NaN          NaN     NaN        NaN    NaN     NaN   
max                  NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand  \
count       50000        50000.000000     45518       50000   
unique         13                 NaN         7          40   
top     150,000km                 NaN    benzin  volkswagen   
freq        32424                 NaN     30107       10687   
mean          NaN            5.723360       NaN         NaN   
std           NaN            3.711984       NaN         NaN   
min           NaN            0.000000       NaN         NaN   
25%           NaN            3.000000       NaN         NaN   
50%           NaN            6.000000       NaN         NaN   
75%           NaN            9.000000       NaN         NaN   
max           NaN           12.000000       NaN         NaN   

       unrepaired_damange      ad_created  nr_of_pictures   postal_code  \
count               40171           50000         50000.0  50000.000000   
unique                  2              76             NaN           NaN   
top                  nein  3/04/2016 0:00             NaN           NaN   
freq                35232            1946             NaN           NaN   
mean                  NaN             NaN             0.0  50813.627300   
std                   NaN             NaN             0.0  25779.747957   
min                   NaN             NaN             0.0   1067.000000   
25%                   NaN             NaN             0.0  30451.000000   
50%                   NaN             NaN             0.0  49577.000000   
75%                   NaN             NaN             0.0  71540.000000   
max                   NaN             NaN             0.0  99998.000000   

             last_seen  
count            50000  
unique           10432  
top     6/04/2016 2:45  
freq               112  
mean               NaN  
std                NaN  
min                NaN  
25%                NaN  
50%                NaN  
75%                NaN  
max                NaN

Seller and offer_type have only 2 unique values one of which appears 4999 times, so these columns can be dropped.

Cols that may need further investigating: 
resgistration_year and registration_month because it is showing NaN for top and unique values and hence unexpected min/max stats. Nr_of_pictures may also need investigating because that column has no stats.

The columns price and odometer need to be converted to numeric.

In [44]:
autos = autos.drop(["seller", "offer_type"], axis=1)

#### Clean 'price' and 'odometer' columns

In [9]:
# Clean price column and convert to float
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "")
autos["price"] = autos["price"].astype("float")
autos["price"].dtype

dtype('float64')

In [10]:
# Clean odeometer column and convert to float
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].str.replace(",", "")
autos["odometer"] = autos["odometer"].astype("float")
autos["odometer"].dtype

dtype('float64')

In [11]:
# Rename odometer column to indicate its units.
autos = autos.rename(columns={"odometer":"odometer_km"})
autos.head()

date_crawled                                               name  \
0  26/03/2016 17:47                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   4/04/2016 13:38         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  26/03/2016 18:57                         Volkswagen_Golf_1.6_United   
3  12/03/2016 16:58  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   1/04/2016 14:38  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damange       ad_created  nr_of_pictures  \
0     peugeot               nein  26/03/2016 0:00               0   
1         bmw               nein   4/04/2016 0:00               0   
2  volkswagen               nein  26/03/2016 0:00               0   
3       smart               nein  12/03/2016 0:00               0   
4        ford               nein   1/04/2016 0:00               0   

   postal_code        last_seen  
0        79588   6/04/2016 6:45  
1        71034  6/04/2016 14:45  
2        35394  6/04/2016 20:15  
3        33729  15/03/2016 3:16  
4        39218  1/04/2016 14:38

#### Explore the odometer_km column and clean further if needed.

In [12]:
# Explore odometer_km
autos["odometer_km"].unique().shape

(13,)

In [13]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [14]:
autos["odometer_km"].value_counts()

odometer_km
150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: count, dtype: int64

The column odometer_km is to the nearest 1000 kms and there's only 13 unique values.

#### Explore price column and clean further if needed.

In [15]:
# Explore price data
autos["price"].unique().shape

(2357,)

In [16]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [17]:
print(autos["price"].value_counts().sort_index(ascending=True).head(20))
print(autos["price"].value_counts().sort_index(ascending=True).tail(20))

price
0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: count, dtype: int64
price
197000.0      1
198000.0      1
220000.0      1
250000.0      1
259000.0      1
265000.0      1
295000.0      1
299000.0      1
345000.0      1
350000.0      1
999990.0      1
999999.0      2
1234566.0     1
1300000.0     1
3890000.0     1
10000000.0    1
11111111.0    2
12345678.0    3
27322222.0    1
99999999.0    1
Name: count, dtype: int64


In [18]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

Considering that the cars are for sale, we should exlude cars with a price of nil. We can keep cars with a price 1 dollar and above as ebay is an auction site and it could be that some cars have a starting price of 1 dollar. Considering that the most expensive car brand is Porsche and on average their cars are less than $999,990, we can exclude cars with prices greater than that, as there must be some error in the data.

In [19]:
autos = autos[autos["price"].between(1, 350001)]

In [20]:
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

#### Exploring the date columns

In [45]:
# Distribution of date_crawled
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

date_crawled
1/04/2016     0.03380
10/03/2016    0.03212
11/03/2016    0.03248
12/03/2016    0.03678
13/03/2016    0.01556
14/03/2016    0.03662
15/03/2016    0.03398
16/03/2016    0.02950
17/03/2016    0.03152
18/03/2016    0.01306
19/03/2016    0.03490
2/04/2016     0.03540
20/03/2016    0.03782
21/03/2016    0.03752
22/03/2016    0.03294
23/03/2016    0.03238
24/03/2016    0.02910
25/03/2016    0.03174
26/03/2016    0.03248
27/03/2016    0.03104
28/03/2016    0.03484
29/03/2016    0.03418
3/04/2016     0.03868
30/03/2016    0.03362
31/03/2016    0.03192
4/04/2016     0.03652
5/03/2016     0.02538
5/04/2016     0.01310
6/03/2016     0.01394
6/04/2016     0.00318
7/03/2016     0.03596
7/04/2016     0.00142
8/03/2016     0.03330
9/03/2016     0.03322
Name: proportion, dtype: float64

The date crawled is between 5th March and 7th April of 2016 with approximately 3% of cars crawled each day.

In [46]:
# Distribution of ad_created
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

ad_created
1/02/2016     0.00002
1/03/2016     0.00010
1/04/2016     0.03380
10/01/2016    0.00004
10/03/2016    0.03186
               ...   
8/02/2016     0.00002
8/03/2016     0.03334
9/02/2016     0.00004
9/03/2016     0.03324
9/09/2015     0.00002
Name: proportion, Length: 76, dtype: float64

The ads were created from February 2015 to April 2016. From March 2016, more ads started to be created per day, with percentage jumping from 0.002% to 3%.

In [49]:
# Distribution of last_seen
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

last_seen
1/04/2016     0.02310
10/03/2016    0.01076
11/03/2016    0.01252
12/03/2016    0.02382
13/03/2016    0.00898
14/03/2016    0.01280
15/03/2016    0.01588
16/03/2016    0.01644
17/03/2016    0.02792
18/03/2016    0.00742
19/03/2016    0.01574
2/04/2016     0.02490
20/03/2016    0.02070
21/03/2016    0.02074
22/03/2016    0.02158
23/03/2016    0.01858
24/03/2016    0.01956
25/03/2016    0.01920
26/03/2016    0.01696
27/03/2016    0.01602
28/03/2016    0.02086
29/03/2016    0.02234
3/04/2016     0.02536
30/03/2016    0.02484
31/03/2016    0.02384
4/04/2016     0.02462
5/03/2016     0.00108
5/04/2016     0.12428
6/03/2016     0.00442
6/04/2016     0.22100
7/03/2016     0.00536
7/04/2016     0.13092
8/03/2016     0.00760
9/03/2016     0.00986
Name: proportion, dtype: float64

The ads were last seen for all cars between March and April of 2016. Each day of the distribution constitutes about 2% of all cars.

#### Epxloring the registration_year column and cleaning if needed

In [24]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

On average, cars were registered in 2005. 75% of the cars were registered before 2008. There is at least 2 outliers, as the minimum value is 1000 and the maximum value is 9999, which doesn't make sense.

Cars were only invented after 1900, and the latest year in which the ads were created is 2016. So any registration year outside this doesn't make sense and hence can be removed.

In [52]:
# autos[autos["registration_year"].between(1900, 2016)]["registration_year"].value_counts(normalize=True, dropna=False).sort_index()
autos = autos[autos["registration_year"].between(1900, 2016)]
autos["registration_year"].value_counts(normalize=True, dropna=False).sort_index().tail(30)

registration_year
1987    0.001562
1988    0.002957
1989    0.003769
1990    0.008224
1991    0.007412
1992    0.008141
1993    0.009265
1994    0.013742
1995    0.027338
1996    0.030066
1997    0.042225
1998    0.051074
1999    0.062464
2000    0.069834
2001    0.056280
2002    0.052740
2003    0.056779
2004    0.056988
2005    0.062776
2006    0.056384
2007    0.047972
2008    0.046452
2009    0.043683
2010    0.033251
2011    0.034022
2012    0.027546
2013    0.016782
2014    0.013867
2015    0.008308
2016    0.027401
Name: proportion, dtype: float64

#### Explore price by brand

In [31]:
# Explore brand column
# top20_brands = list(autos["brand"].value_counts(normalize=True, dropna=False).head(20).index)
autos["brand"].value_counts(normalize=True, dropna=False)

brand
volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.00

There are 40 brands; let's aggregate on the top 20 brands and calculate their mean prices.

In [27]:
top20_brands = list(autos["brand"].value_counts(normalize=True, dropna=False).head(20).index)
top20_brands

['volkswagen',
 'bmw',
 'opel',
 'mercedes_benz',
 'audi',
 'ford',
 'renault',
 'peugeot',
 'fiat',
 'seat',
 'skoda',
 'nissan',
 'mazda',
 'smart',
 'citroen',
 'toyota',
 'hyundai',
 'sonstige_autos',
 'volvo',
 'mini']

In [28]:
mean_price_dict = {}
for brand in top20_brands:
    selected_rows = autos[autos["brand"] == brand]
    mean = (selected_rows["price"].sum()) / len(selected_rows["price"])
    mean_price_dict[brand] = mean

In [29]:
pd.Series(mean_price_dict).sort_values(ascending=False)

sonstige_autos    12338.550218
mini              10613.459658
audi               9336.687454
mercedes_benz      8628.450366
bmw                8332.820518
skoda              6368.000000
volkswagen         5402.410262
hyundai            5365.254274
toyota             5167.091062
volvo              4946.501171
nissan             4743.402525
seat               4397.230950
mazda              4112.596615
citroen            3779.139144
ford               3749.469507
smart              3580.223903
peugeot            3094.017229
opel               2975.241935
fiat               2813.748538
renault            2474.864607
dtype: float64

Within the the top 20 brands, brand with highest mean price is Sonstige Autos and lowest mean price is Renault.

What if we wanted to investigate the only the first few brands? There are only 6 brands with listings that make more than 5% of the total listings so let's aggregate on them. Let's simultaneously investigate how mileage varies between these brands, and whether there is a relationship between price and mileage.

#### Exploring price and mileage by brand

In [32]:
top6_mean_mileage_dict = {}
top6_mean_price_dict = {}
for brand in top20_brands[:6]:
    selected_rows = autos[autos["brand"] == brand]
    mean_price = (selected_rows["price"].sum()) / len(selected_rows["price"])
    top6_mean_price_dict[brand] = mean_price
    mean_mileage = (selected_rows["odometer_km"].sum()) / len(selected_rows["odometer_km"])
    top6_mean_mileage_dict[brand] = mean_mileage    

In [51]:
mean_mileage_series = pd.Series(top6_mean_mileage_dict)
print(mean_mileage_series)

volkswagen       128955.272761
opel             129298.663248
bmw              132521.643028
mercedes_benz    130886.142797
audi             129643.941163
ford             124131.934464
dtype: float64


In [52]:
mean_price_series = pd.Series(top6_mean_price_dict)
print(mean_price_series)

volkswagen        6384.167400
opel              5106.092657
bmw               8252.918954
mercedes_benz    29511.955429
audi              8965.560355
ford              7105.662547
dtype: float64


In [53]:
top6_df = pd.DataFrame(mean_price_series, columns=["mean_price"])
top6_df

mean_price
volkswagen      6384.167400
opel            5106.092657
bmw             8252.918954
mercedes_benz  29511.955429
audi            8965.560355
ford            7105.662547

In [54]:
top6_df["mean_mileage"] = mean_mileage_series
top6_df

mean_price   mean_mileage
volkswagen      6384.167400  128955.272761
opel            5106.092657  129298.663248
bmw             8252.918954  132521.643028
mercedes_benz  29511.955429  130886.142797
audi            8965.560355  129643.941163
ford            7105.662547  124131.934464

In [55]:
top6_df.sort_values(by="mean_price", axis=0)

mean_price   mean_mileage
opel            5106.092657  129298.663248
volkswagen      6384.167400  128955.272761
ford            7105.662547  124131.934464
bmw             8252.918954  132521.643028
audi            8965.560355  129643.941163
mercedes_benz  29511.955429  130886.142797

The brand with the highest mean price is Mercedes-Benz and the lowest is Opel. The brand with the highest mean mileage is BMW and the lowest is Ford. 

Generally, one would expect that as mileage decreases, the price would be higher. We can see from the above that in the luxury cars sector (i.e. the three most expensive brands), this does not hold true, as Mercedes Benz despite having a higher mileage is still higher in price than Audi. It shows that customers in this market segment value the brand more than the mileage in determining what price they're willing to pay. Conversely, in the middle of the market (i.e. next 3 expensive brands), the general relationship between price and mileage holds true. 